### PR TITLE
Add page on symbols 

### PR DESCRIPTION
**Pull request by sywhang**
## Summary

Add a page on symbols by moving the short paragraph description we had, and added some contents. The added content was mainly migrated from  https://github.com/dotnet/core/blob/master/Documentation/diagnostics/portable_pdb.md.

Part of https://github.com/dotnet/diagnostics/issues/515

### CODE REVIEW

**Review by noahfalk — APPROVED** [review]
I suggested a few edits inline but mostly I am hoping we can get @tmat our portable pdb expert to take a look : )

File: docs/core/diagnostics/symbols.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+---
+title: Symbols in .NET
+description: An introduction to symbols and portable PDBs in .NET
+ms.date: 02/08/2021
+---
+
+# Symbols
+
+Symbols are useful for debugging and other diagnostic tools. The contents of symbol files vary between languages, compilers, and platforms. At a very high level symbols are a mapping between the source code and the binary produced by the compiler. These mappings are used by tools like [Visual Studio](/visualstudio/debugger/what-is-debugging) and [Visual Studio Code](https://code.visualstudio.com/Docs/editor/debugging) to resolve source line number information or local variable names.
```
Comment (Youssef1313) [inline docs/core/diagnostics/symbols.md:9]:
```suggestion
Symbols are useful for debugging and other diagnostic tools. The contents of symbol files vary between languages, compilers, and platforms. At a very high level, symbols are a mapping between the source code and the binary produced by the compiler. These mappings are used by tools like [Visual Studio](/visualstudio/debugger/what-is-debugging) and [Visual Studio Code](https://code.visualstudio.com/Docs/editor/debugging) to resolve source line number information or local variable names.
```

File: docs/core/diagnostics/symbols.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+The [Windows documentation on symbols](/windows/win32/dxtecharts/debugging-with-symbols) contain more detailed information on symbols for Windows, although many of the concepts apply to other platforms as well.
+
+## Learn about .NET's Portable PDB format
```
Comment (Youssef1313) [inline docs/core/diagnostics/symbols.md:13]:
```suggestion
## .NET's Portable PDB (Program Database) format
```

File: docs/core/diagnostics/symbols.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+## Use the correct PDB format for your scenario
+
+Neither portable PDBs nor Windows PDBs are supported everywhere, so you need to consider where your project will want to be used and debugged to decide which format to use. If you have a project that you want to be able to use and debug in both formats, you can use different build configurations and build the project twice to support both types of consumer.
+
+### Support for Windows PDBs
+
+Windows PDBs can only be written or read on Windows. All Windows tooling supports them, except for Visual Studio Code (as Visual Studio Code strives for consistent behavior across all platforms), and scenarios where Visual Studio is debugging to a remote Linux/OSX computer (as the PDBs must be read on the remote computer).
```
Comment (Youssef1313) [inline docs/core/diagnostics/symbols.md:27]:
```suggestion
Windows PDBs can only be written or read on Windows. All Windows tooling supports them, except for Visual Studio Code (as Visual Studio Code strives for consistent behavior across all platforms), and scenarios where Visual Studio is debugging to a remote Linux/macOS computer (as the PDBs must be read on the remote computer).
```

File: docs/core/diagnostics/symbols.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+### Support for Portable PDBs
+
+Portable PDBs can be read on any operating system, but there are a number of places where they aren't supported yet:
+
+* Older versions of the Visual Studio debugger (versions before VS 2015 Update 2)
+
+* Applications targeting .NET Framework 4.7.1 or earlie: printing stack traces with mappings back to line numbers (such as in an ASP.NET error page). The name of methods is unaffected, only the source file names and line numbers are unsupported
```
Comment (Youssef1313) [inline docs/core/diagnostics/symbols.md:35]:
```suggestion
* Applications targeting .NET Framework 4.7.1 or earlier: printing stack traces with mappings back to line numbers (such as in an ASP.NET error page). The name of methods is unaffected, only the source file names and line numbers are unsupported
```

File: docs/core/diagnostics/symbols.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+### What is a PDB?
+
+A PDB file is an auxiliary file produced by a compiler to provide other tools, especially debuggers, information about what is in the main executable file and how it was produced. For example, a debugger reads a PDB to map foo.cs line 12 to the right executable location so that it can set a breakpoint. The Windows PDB format has been around a long time, and it evolved from other native debugging symbol formats which were even older. It started out its life as a format for native (C/C++) programs. For the first release of the .NET Framework, the Windows PDB format was extended to support .NET.
+
+## Use the correct PDB format for your scenario
+
+Neither portable PDBs nor Windows PDBs are supported everywhere, so you need to consider where your project will want to be used and debugged to decide which format to use. If you have a project that you want to be able to use and debug in both formats, you can use different build configurations and build the project twice to support both types of consumer.
```
[the remark this answers is not in the document]
Comment (sywhang) [reply docs/core/diagnostics/symbols.md:23]: @tmat Do you have any recommendations on the questions above? Thanks in advance!

File: docs/core/diagnostics/symbols.md (hunk 1 above)
Comment (tmat) [reply docs/core/diagnostics/symbols.md:9]: Mention [WinDBG](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools)?

File: docs/core/diagnostics/symbols.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (tmat) [reply docs/core/diagnostics/symbols.md:23]: Agree with @noahfalk. I think we should go even further and say that using Windows PDBs for managed code is obsolete and only needed for legacy tools. We should recommend using Portable PDBs for all projects (whether they target .NET Core or .NET Framework). There are some newer compiler features are only implemented for Portable PDBs.

DIA, WinBDG now support Portable PDB. There might be some profilers that still do not support Portable PDBs. For these cases I'd point to [Pdb2Pdb converter](https://github.com/dotnet/symreader-converter#pdb2pdb).

[the remark this answers is not in the document]
Comment (tmat) [reply docs/core/diagnostics/symbols.md:23]: NuGet.org does not support Windows PDBs in snupkg.

File: docs/core/diagnostics/symbols.md (hunk 3 above)
Comment (tmat) [reply docs/core/diagnostics/symbols.md:27]: Some tools do not support Windows PDBs. E.g. publishing non-Portable PDBs in snupkg to Nuget.org is not supported. Some upcoming tools also will not support Windows PDBs.

File: docs/core/diagnostics/symbols.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+### Support for Portable PDBs
+
+Portable PDBs can be read on any operating system, but there are a number of places where they aren't supported yet:
+
+* Older versions of the Visual Studio debugger (versions before VS 2015 Update 2)
+
+* Applications targeting .NET Framework 4.7.1 or earlie: printing stack traces with mappings back to line numbers (such as in an ASP.NET error page). The name of methods is unaffected, only the source file names and line numbers are unsupported
+
+* C# Code analysis tools (such as FxCop) except Roslyn Analyzer
```
Comment (tmat) [inline docs/core/diagnostics/symbols.md:37]: FxCop does not support Portable PDBs but it is now deprecated. Roslyn Analyzers are the replacement.

File: docs/core/diagnostics/symbols.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+* Some symbol servers (for example, [SymbolSource](http://www.symbolsource.org) does not support portable PDBs, but [NuGet](https://nuget.org) does)
+
+* Running post-compilation build step that consumes or modifies the PDB using older versions of tools such as CCI, CodeContracts
+
+* Using .NET decompilers such as ildasm or .NET reflector and expecting to see source line mappings or local parameter names
```
Comment (tmat) [inline docs/core/diagnostics/symbols.md:43]: I think the latest ildasm now supports Portable PDBs but am not 100% sure. @noahfalk?
ILSpy decompiler supports Portable PDBs. IDK about .NET Reflector. Not sure if it's important.

File: docs/core/diagnostics/symbols.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+* MS DIA-based tools such as WinDBG.
```
Comment (tmat) [inline docs/core/diagnostics/symbols.md:45]: DIA and WinDBG now support Portable PDB. I'd say "The latest versions of DIA and tools using it for reading symbols, such as WinDBG support Portable PDBs, but older version do not.".

Comment (tmat) [reply docs/core/diagnostics/symbols.md:45]: Link to DIA: https://docs.microsoft.com/en-us/visualstudio/debugger/debug-interface-access/debug-interface-access-sdk?view=vs-2019

File: docs/core/diagnostics/symbols.md (hunk 3 above)
Comment (tmat) [reply docs/core/diagnostics/symbols.md:27]: /cc @clairernovotny Let's make sure this info is aligned with our .NET ecosystem goals and "verifiable" packages.

**Comment by sywhang** [comment]
Thanks for the feedback @tmat @noahfalk -- I updated the texts to be more bullish towards portable PDBs and worded Windows PDB as legacy. Could you PTAL when you get a chance?

**Comment by tmat** [comment]
> worded  Windows PDB as legacy

Let's emphasize that this is Windows PDBs for **managed code** that are legacy. It does not apply to Windows PDBs for native (C++, ngen, crossgen).

**Comment by sywhang** [comment]
@tmat Yup -- Sorry should've made my comment a little more clear. The docs change I made emphasizes that point for managed code only:

```
Using Windows PDBs for managed code is obsolete and is only needed for legacy tools. 
```

File: docs/core/diagnostics/symbols.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (noahfalk) [reply docs/core/diagnostics/symbols.md:23]: > WinBDG now support Portable PDB

Not fully : ) @davmason is still working with the windbg team to get to a better place on portable PDB support. The lack of this support is one of the reasons that .NET still produces and indexes windows PDBs.

**Pull request merged**